src/bitmap_font.h

```diff
@@ -0,0 +1,224 @@
+#pragma once
+
+#include "global.h"
+#include "graphics.h"
+#include "data/bitmap_font_data.h"
+
+// Definition of a Bitmap Font.
+struct BitmapFontData {
+    u8 whitespaceWidth;  // Whitespace Width
+    u8 textureWidth;     // Glyph Texture Width
+    u8 textureHeight;    // Glyph Texture Height
+    u8 descensionHeight; // Descension Height for Lowercase Latin Alphabet Characters
+    s8 spacingWidth;     // Glyph Spacing
+    // [(0x8140 .. 0x817E), (0x8180 .. 0x819E)]
+    const void *punctuationTextures;
+    const u8 *punctuationWidths;
+    // [(0x824F .. 0x8258)]
+    const void *arabicNumeralTextures;
+    const u8 *arabicNumeralWidths;
+    // [(0x8260 .. 0x8279)]
+    const void *latinUppercaseTextures;
+    const u8 *latinUppercaseWidths;
+    // [(0x8281 .. 0x829A)]
+    const void *latinLowercaseTextures;
+    const u8 *latinLowercaseWidths;
+    // [(0x829F .. 0x82FC)]
+    const void *hiraganaTextures;
+    const u8 *hiraganaWidths;
+    // [(0x8340 .. 0x837E), (0x8380 .. 0x839E)]
+    const void *katakanaTextures;
+    const u8 *katakanaWidths;
+};
+
+// Bitmap Font Printer which creates OBJ Animations.
+struct BitmapFontOBJ {
+    u16 memID;
+    const struct BitmapFontData *fonts;
+    u16 baseTileNum;
+    u8 maxAllocatedTileRows;
+    u16 *printedGlyphs;
+    u8 *printedGlyphCounts;
+    void (*parseString)(char *, const char *);
+    char *parsedOutput;
+};
+
+// Bitmap Font Printer which writes to BG Maps.
+struct BitmapFontBG {
+    const struct BitmapFontData *fonts;
+    u16 baseTileNum;
+    u8 tilesetID;
+    u8 maxAllocatedTileRows;
+    u16 *printedGlyphs;
+    u8 *printedGlyphCounts;
+};
+
+// Generated OBJ Animation for text.
+struct PrintedTextAnim {
+    struct Animation frames[2];
+    u16 oam[0];
+};
+
+// Generated OBJ Animation for wobbly/shaky text.
+struct WobblyPrintedTextAnim {
+    struct Animation frames[6];
+    u16 oam[0];
+};
+
+// Text queue for the multi-string BitmapFontOBJ printer task.
+struct BitmapFontQueuedText {
+    const char *string;
+    u32 alignment:4;
+    u32 fontStyle:4;
+    u32 palette:4;
+};
+
+// Multi-string BitmapFontOBJ printer task.
+struct BitmapFontOBJPrinter {
+    struct BitmapFontOBJ *textObj;
+    struct PrintedTextAnim **animTable;
+    struct BitmapFontQueuedText *queue;
+    u32 current;
+};
+
+// BitmapFontBG printer task.
+struct BitmapFontBGPrinter {
+    struct BitmapFontBG *textObj;
+    u32 fontStyle:4;
+    u32 palette:4;
+    u32 processLimit:8;
+    u32 currentMapX:10;
+    u16 bgMapWidth;
+    u16 *bgMap;
+    const char *string;
+};
+
+
+extern void bmp_font_get_glyph(const struct BitmapFontData *font, const char *string, const u16 **textureReq, u32 *widthReq); // Get printable glyph texture and width.
```

Comment:
i don't like how clogged having these comments in with the header looks. these can just stay above the function's real definition, i think most code editors will still pick it up like that. these should just stay uncommented imo